00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,377 @@
+
+// Napi version of node::MakeCallback(...)
+NODE_EXTERN napi_value napi_make_callback(napi_env e, napi_value recv,
+                                          napi_value func,
+                                          int argc, napi_value* argv);
+
+
+// Methods to work with napi_callbacks
+NODE_EXTERN int napi_get_cb_args_length(napi_env e, napi_func_cb_info cbinfo);
+NODE_EXTERN void napi_get_cb_args(napi_env e, napi_func_cb_info cbinfo,
+                                  napi_value* buffer, size_t bufferlength);
+NODE_EXTERN napi_value napi_get_cb_this(napi_env e, napi_func_cb_info cbinfo);
+// V8 concept; see note in .cc file
+NODE_EXTERN napi_value napi_get_cb_holder(napi_env e, napi_func_cb_info cbinfo);
+NODE_EXTERN bool napi_is_construct_call(napi_env e, napi_func_cb_info cbinfo);
+NODE_EXTERN void napi_set_return_value(napi_env e, napi_func_cb_info cbinfo,
+                                       napi_value v);
+
+
+// Methods to support ObjectWrap
+NODE_EXTERN napi_value napi_create_constructor_for_wrap(napi_env e,
+                                                        napi_callback cb);
+NODE_EXTERN void napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
+                           napi_destruct* napi_destructor,
+                           napi_persistent* handle);
+NODE_EXTERN void* napi_unwrap(napi_env e, napi_value jsObject);
+
+NODE_EXTERN napi_value napi_create_constructor_for_wrap_with_methods(
+  napi_env e,
+  napi_callback cb,
+  char* utf8name,
+  int methodcount,
+  napi_method_descriptor* methods);
+
+// Methods to control object lifespan
+NODE_EXTERN napi_persistent napi_create_persistent(napi_env e, napi_value v);
+NODE_EXTERN void napi_release_persistent(napi_env e, napi_persistent p);
+NODE_EXTERN napi_value napi_get_persistent_value(napi_env e, napi_persistent p);
+NODE_EXTERN napi_handle_scope napi_open_handle_scope(napi_env e);
+NODE_EXTERN void napi_close_handle_scope(napi_env e, napi_handle_scope s);
+NODE_EXTERN napi_escapable_handle_scope
+                napi_open_escapable_handle_scope(napi_env e);
+NODE_EXTERN void napi_close_escapable_handle_scope(
+                                             napi_env e,
+                                             napi_escapable_handle_scope s);
+NODE_EXTERN napi_value napi_escape_handle(napi_env e,
+                                          napi_escapable_handle_scope s,
+                                          napi_value v);
+
+
+// Methods to support error handling
+NODE_EXTERN void napi_throw(napi_env e, napi_value error);
+NODE_EXTERN void napi_throw_error(napi_env e, char* msg);
+NODE_EXTERN void napi_throw_type_error(napi_env e, char* msg);
+
+// These are particularly rough, needs more analysis
+typedef void (*napi_try_callback)(napi_env e, void* data);
+typedef void (*napi_catch_callback)(napi_env e, void* data);
+NODE_EXTERN bool napi_try_catch(napi_env e, napi_try_callback t,
+                                napi_catch_callback c, void* data);
```

Comment:
Create https://github.com/nodejs/abi-stable-node/issues/27 to discuss as part of prototype